File: sample-objc/ARGEARSample/UI/UIComponent/RatioView.m
//
//  RatioView.m
//  ARGEARSample
//
//  Created by Jihye on 10/09/2019.
//  Copyright © 2019 Seerslab. All rights reserved.
//

#import "RatioView.h"

#define kRatioViewTopBottomAlign169 -44
#define kRatioViewTopBottomAlign43 0
#define kRatioViewTopBottomAlign11 64

#define kRatioViewBottomViewTopAlign169 300
#define kRatioViewBottomViewTopAlign43 0
#define kRatioViewBottomViewTopAlign11 -([[UIScreen mainScreen] bounds].size.width * 4/3 - [[UIScreen mainScreen] bounds].size.width) + 64

@implementation RatioView

/*
// Only override drawRect: if you perform custom drawing.
// An empty implementation adversely affects performance during animation.
- (void)drawRect:(CGRect)rect {
    // Drawing code
}
*/

- (void)awakeFromNib {
    [super awakeFromNib];
    
    [self setRatio:ARGMediaRatio_4x3];
}

- (void)setRatio:(ARGMediaRatio)ratio {
    
    CGFloat screenRatio = [[UIScreen mainScreen] bounds].size.height / [[UIScreen mainScreen] bounds].size.width;
    
    switch (ratio) {
        case ARGMediaRatio_16x9:
            if (screenRatio > 2.f) {
                [_topViewBottomAlign setConstant:kRatioViewTopBottomAlign169];
            } else {
                [_topViewBottomAlign setConstant:0];
            }
            [_bottomView setHidden:YES];
            break;
        case ARGMediaRatio_4x3:
            if (screenRatio > 2.f) {
                [_topViewBottomAlign setConstant:kRatioViewTopBottomAlign11];
                [_bottomViewTopAlign setConstant:kRatioViewBottomViewTopAlign43 + 64.f];
            } else {
                [_topViewBottomAlign setConstant:kRatioViewTopBottomAlign43];
                [_bottomViewTopAlign setConstant:kRatioViewBottomViewTopAlign43];
            }
            [_bottomView setHidden:NO];
            break;
        case ARGMediaRatio_1x1:
            [_topViewBottomAlign setConstant:kRatioViewTopBottomAlign11];
            [_bottomViewTopAlign setConstant:kRatioViewBottomViewTopAlign11];
            [_bottomView setHidden:NO];
            break;
        default:
            break;
    }
}

@end

File: sample-swift/ARGearSample/UI/UIComponent/RatioView.swift
//
//  RatioView.swift
//  ARGearSample
//
//  Created by Jihye on 2020/01/30.
//  Copyright © 2020 Seerslab. All rights reserved.
//

import UIKit
import ARGear

let kRatioViewTopBottomAlign169 = -44
let kRatioViewTopBottomAlign43 = 0
let kRatioViewTopBottomAlign11 = 64

let kRatioViewBottomViewTopAlign169 = 260
let kRatioViewBottomViewTopAlign43 = 0
let kRatioViewBottomViewTopAlign11 = -(UIScreen.main.bounds.size.width * 4/3 - UIScreen.main.bounds.size.width) + 64

typealias RatioViewCompletion = () -> Void

class RatioView: UIView {
    @IBOutlet weak var topView: UIView!
    @IBOutlet weak var bottomView: UIView!
    @IBOutlet weak var blurView: UIView!
    
    @IBOutlet weak var topViewBottomAlign: NSLayoutConstraint!
    @IBOutlet weak var bottomViewTopAlign: NSLayoutConstraint!
    
    override func awakeFromNib() {
        super.awakeFromNib()
        
        self.setRatio(ARGMediaRatio(rawValue: self.tag) ?? ._4x3)
    }
    
    func setRatio(_ ratio: ARGMediaRatio) {
        
        let screenRatio = UIScreen.main.bounds.size.height / UIScreen.main.bounds.size.width
        switch ratio {
        case ._16x9:
            var topViewBottomAlignConstant = 0
            if screenRatio > 2.0 {
                topViewBottomAlignConstant = kRatioViewTopBottomAlign169
            }
            self.animate({
                self.topViewBottomAlign.constant = CGFloat(topViewBottomAlignConstant)
                self.bottomViewTopAlign.constant = CGFloat(kRatioViewBottomViewTopAlign169)
            }) {
                self.bottomView.isHidden = true
            }
        case ._4x3:
            var topViewBottomAlignConstant = kRatioViewTopBottomAlign43
            var bottomViewTopAlignConstant = kRatioViewBottomViewTopAlign43
            if screenRatio > 2.0 {
                topViewBottomAlignConstant = kRatioViewTopBottomAlign11
                bottomViewTopAlignConstant = kRatioViewBottomViewTopAlign43 + 64
            }
            
            self.bottomView.isHidden = false
            self.animate({
                self.topViewBottomAlign.constant = CGFloat(topViewBottomAlignConstant)
                self.bottomViewTopAlign.constant = CGFloat(bottomViewTopAlignConstant)
            }, completion: nil)
        case ._1x1:
            self.bottomView.isHidden = false
            self.animate({
                self.topViewBottomAlign.constant = CGFloat(kRatioViewTopBottomAlign11)
                self.bottomViewTopAlign.constant = kRatioViewBottomViewTopAlign11
            }, completion: nil)
        default:
            break
        }
    }
    
    func blur(_ blur: Bool) {
        self.blurView.isHidden = !blur
    }
    
    private func animate(_ animate: @escaping RatioViewCompletion, completion: RatioViewCompletion?) {
        DispatchQueue.main.async {
            UIView.animate(withDuration: 0.2, delay: 0, options: .curveEaseIn, animations: {
                animate()
                self.layoutIfNeeded()
            }) { complete in
                completion?()
                self.blur(false)
            }
        }
    }
}
